lib/Basics/FutureSharedLock.h

```diff
@@ -0,0 +1,244 @@
+
+      auto it = insertNode(true);
+      blockedFunc(it);
+      return (*it)->promise.getFuture();
+    }
+
+    template<class Func>
+    FutureType asyncLockShared(Func blockedFunc) {
+      std::lock_guard lock(_mutex);
+      if (_lockCount == 0 || (!_exclusive && _queue.empty())) {
+        ++_lockCount;
+        _exclusive = false;
+        return {LockGuard(this)};
+      }
+
+      auto it = insertNode(false);
+      blockedFunc(it);
+      return (*it)->promise.getFuture();
+    }
+
+    void unlock() {
+      std::lock_guard lock(_mutex);
+      TRI_ASSERT(_lockCount > 0);
+      if (--_lockCount == 0 && !_queue.empty()) {
+        // we were the last lock holder -> schedule the next node
+        auto& node = _queue.back();
+        _lockCount = 1;
+        _exclusive = node->exclusive;
+        scheduleNode(node);
+        _queue.pop_back();
+
+        // if we are in shared mode, we can schedule all following shared
+        // nodes
+        if (!_exclusive) {
+          while (!_queue.empty() && !_queue.back()->exclusive) {
+            ++_lockCount;
+            scheduleNode(_queue.back());
+            _queue.pop_back();
+          }
+        }
+      }
+    }
+
+    auto insertNode(bool exclusive) {
+      _queue.emplace_front(std::make_shared<Node>(exclusive));
+      return _queue.begin();
+    }
+
+    void removeNode(
+        typename std::list<std::shared_ptr<Node>>::iterator queueIterator) {
+      TRI_ASSERT(!_queue.empty());
+      TRI_ASSERT(_lockCount > 0);
+      if (std::addressof(_queue.back()) == std::addressof(*queueIterator)) {
+        // we are the first entry in the list -> remove ourselves and check if
+        // we can schedule the next node(s)
+        _queue.pop_back();
+        if (!_exclusive) {
+          while (!_queue.empty() && !_queue.back()->exclusive) {
+            ++_lockCount;
+            scheduleNode(_queue.back());
+            _queue.pop_back();
+          }
+        }
+      } else {
+        // we are somewhere in the middle -> just remove ourselves
+        _queue.erase(queueIterator);
+      }
+    }
+
+    void scheduleExclusive(Node* node) { scheduleNode(node); }
+
+    void scheduleShared(Node* leader) { scheduleNode(leader); }
+
+    void scheduleNode(std::shared_ptr<Node> node) {
+      TRI_ASSERT(node != nullptr);
+      node->scheduled = true;
+      _scheduler.queue(
+          [node, this]() { node->promise.setValue(LockGuard(this)); });
+    }
+
+    void scheduleTimeout(
+        typename std::list<std::shared_ptr<Node>>::iterator queueIterator,
+        std::chrono::milliseconds timeout) {
+      _scheduler.queueDelayed(
+          [self = this->weak_from_this(),
+           node = std::weak_ptr<Node>(*queueIterator),
+           queueIterator]() mutable {
+            auto me = self.lock();
+            if (me) {
+              std::lock_guard lock(me->_mutex);
+              auto nodePtr = node.lock();
+              if (nodePtr != nullptr && !nodePtr->scheduled) {
```

Comment:
If we remove storing shared_ptr to scheduleNode callback
we can write something like this
```suggestion
              auto nodePtr = node.lock();
              if (nodePtr == nullptr) {
                return;  // most not expired will be here
              }
              std::lock_guard lock(me->_mutex);
              // we don't use use_count for sync, only as counter under mutex
              if (nodePtr.use_count() != 1) {
```

Reply:
I am not sure I fully understand - the suggested change by itself would be incorrect. But this is the timeout handling, which is by definition not "popular", so it doesn't really make sense to optimize it.

Reply:
This suggestion with correct with this one https://github.com/arangodb/arangodb/pull/18351#discussion_r1390941300

Reply:
And it's kind of popular because it's for every trylock with timeout

Reply:
> If we remove storing shared_ptr to scheduleNode callback

Reply:
Ok, now I think I understand what you mean.